### PR TITLE
Add datasets for human mesh

### PR DESCRIPTION
**Pull request by zengwang430521**
Rebase from new master branch. 
Add the codes of datasets for human mesh estimation.
Add test codes and test data.

### CODE REVIEW

**Comment by innerlee** [comment]
The merge path is a little bit complex, could you simplify it?

Currently:
* `human_mesh_dataset` -> `human_mesh`
* `human_mesh_pipeline` -> `human_mesh_dataset`

and `human_mesh` is not visible yet.

This will make things very hard to manage.

Pls try to pr against master. e.g. if `human_mesh_pipeline` is relatively independent, then it can be merged into master first.
Then `human_mesh_dataset` can be merged into master.
and then `human_mesh`, etc.

The other way, namely everything merges into `human_mesh`, and then merge everything to master, is hard to maintain.

**Comment by innerlee** [comment]
need to rebase master

File: mmpose/core/evaluation/mesh_eval.py [hunk 1]
```diff
@@ -12,14 +12,11 @@ def compute_similarity_transform(source_points, target_points):
     is an 3x3 rotation matrix, t 3x1 translation, s scale. And return the
     transformed 3D points source_points_hat (N x 3). i.e. solves the orthogonal
     Procrutes problem.
-
```
Comment (innerlee) [inline mmpose/core/evaluation/mesh_eval.py:15]: this change can be dropped

Comment (zengwang430521) [reply mmpose/core/evaluation/mesh_eval.py]: OK

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    return a dict containing data from two datasets.
+
+    Args:
+        train_dataset (Dataset): Dataset for 3D human mesh estimation.
+        adv_dataset (Dataset): Dataset for adversarial learning,
+        provides real SMPL parameters.
+    """
+
+    def __init__(self, train_dataset, adv_dataset):
+        super(MeshAdvDataset, self).__init__()
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:22]: `super().__init__()`

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    def __getitem__(self, i):
+        """Given index, get the data from train dataset and randomly sample an
+        item from adversarial dataset.
+
+        Return a dict containing data from train and adversarial dataset.
+        """
+        data = self.train_dataset[i]
+        ind_adv = np.random.randint(
+            low=0, high=len(self.adv_dataset) - 0.5, dtype=np.int)
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:40]: double check that `high=len(self.adv_dataset) - 0.5` is not a bug.

randint accept high as an integer, so `len(self.adv_dataset) - 0.5` will be converted to int first, which means that you will get `len(self.adv_dataset) - 1` in the end. note that `high` is exclusive. so the range is actually from 0 to len(self.adv_dataset) - 2 inclusive

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        data.update(self.adv_dataset[ind_adv % len(self.adv_dataset)])
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:41]: is `data` a dictionary type?

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        self.pipeline = Compose(self.pipeline)
+
+        # flip_pairs
+        # For all mesh dataset, we use 24 joints as CMR and SPIN.
+        self.ann_info['flip_pairs'] = [[0, 5], [1, 4], [2, 3], [6, 11],
+                                       [7, 10], [8, 9], [20, 21], [22, 23]]
+        self.ann_info['use_different_joint_weights'] = False
+        self.ann_info['joint_weights'] = \
+            np.ones(24, dtype=np.float32).reshape(
+                (self.ann_info['num_joints'], 1))
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_base_dataset.py:59]: `.reshape(-1, 1)`

Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_base_dataset.py:59]: assert self.ann_info['num_joints'] == 24

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        self.ann_info['uv_type'] = data_cfg['uv_type']
+        self.ann_info['use_IUV'] = data_cfg['use_IUV']
+        self.iuv_prefix = os.path.join(
+            self.img_prefix, '{}_IUV_gt'.format(self.ann_info['uv_type']))
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_base_dataset.py:64]: f-string

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            has_smpl = 1
+        else:
+            _poses = np.zeros((dataset_len, 72), dtype=np.float32)
+            _betas = np.zeros((dataset_len, 10), dtype=np.float32)
+            has_smpl = 0
+
+        # Get gender data, if available
+        if 'gender' in data.keys():
+            _genders = data['gender']
+            _genders = np.array([0 if str(g) == 'm' else 1
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_base_dataset.py:117]:
```
_genders = np.array([str(g) != 'm' for g in _genders]).astype(np.int32)
```

File: mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        ann_file (str): Path to the annotation file.
+        img_prefix (str): Path to a directory where images are held.
+            Default: None.
+        data_cfg (dict): config
+        pipeline (list[dict | callable]): A sequence of data transforms.
+        test_mode (bool): Store True when building test or
+            validation dataset. Default: False.
+    """
+
+    def evaluate(self, outputs, res_folder, metric='joint_error', **kwargs):
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py:30]: where is this `kwargs` used?

File: mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        pred_joints_3d = pred_joints_3d[joint_mapper, :]
+        pred_pelvis = (pred_joints_3d[[2]] + pred_joints_3d[[3]]) / 2
+        pred_joints_3d = pred_joints_3d - pred_pelvis
+
+        gt_joints_3d = joints_3d[joint_mapper, :]
+        gt_pelvis = (gt_joints_3d[[2]] + gt_joints_3d[[3]]) / 2
+        gt_joints_3d = gt_joints_3d - gt_pelvis
+
+        error = pred_joints_3d - gt_joints_3d
+        error = np.sqrt((error**2).sum(axis=-1)).mean(axis=-1)
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py:102]: `numpy.linalg.norm`?

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    """Mix Dataset for the adversarial training in 3D human mesh estimation
+    task.
+
+    The dataset combines data from two datasets and
+    return a dict containing data from two datasets.
+
+    Args:
+        train_dataset (Dataset): Dataset for 3D human mesh estimation.
+        adv_dataset (Dataset): Dataset for adversarial learning,
+        provides real SMPL parameters.
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:18]: indent

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py (hunk 7 above)
Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_base_dataset.py:117]: btw why gender has the type of int32?

File: mmpose/datasets/datasets/mesh/mosh_dataset.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        return gt_db
+
+    def __len__(self, ):
+        """Get the size of the dataset."""
+        return len(self.db)
+
+    def __getitem__(self, idx):
+        """Get the sample given index."""
+        item = cp.deepcopy(self.db[idx])
+        trival, pose, beta = \
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mosh_dataset.py:60]: does trival means tri-value or a typo of trivial

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py (hunk 3 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:40]: Oh, I will fix this bug.

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py (hunk 4 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:41]: Yes, it's a dict

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py (hunk 7 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_base_dataset.py:117]: In some datasets, there is no gender label, and we use `-1` to represent the unknown gender. So gender is not a bool, and I choose int32 for simplicity.

Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_base_dataset.py:117]: what about np.int vs np.int32?

File: mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py (hunk 8 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py:30]: I just follow the style of top_down_dataset and did not remove this variable. I will remove it.

File: mmpose/datasets/datasets/mesh/mosh_dataset.py (hunk 11 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mosh_dataset.py:60]: Sorry, it's a typo.

File: mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py (hunk 8 above)
Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py:30]: this will cause trouble when the user mistype metric as medic.
because no warning/error will be raised.

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py (hunk 7 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_base_dataset.py:117]: It's ok, I will change to np.int.

File: mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py (hunk 8 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_h36m_dataset.py:30]: OK, I will remove the kwargs.

File: mmpose/datasets/datasets/__init__.py [hunk 12]
```diff
@@ -1,13 +1,14 @@
 from .bottom_up import BottomUpCocoDataset
+from .mesh import MeshAdvDataset, MeshH36MDataset, MeshMixDataset, MoshDataset
 from .top_down import (TopDownAicDataset, TopDownCocoDataset,
                        TopDownCocoWholeBodyDataset, TopDownFreiHandDataset,
                        TopDownMpiiDataset, TopDownMpiiTrbDataset,
-                       TopDownOCHumanDataset, TopDownOneHand10KDataset,
-                       TopDownPanopticDataset)
+                       TopDownOCHumanDataset, TopDownOneHand10KDataset)
```
Comment (innerlee) [inline mmpose/datasets/datasets/__init__.py:6]: TopDownPanopticDataset was accidently removed

Comment (zengwang430521) [reply mmpose/datasets/datasets/__init__.py:6]: I will fix this bug.

File: mmpose/datasets/datasets/mesh/mesh_adv_dataset.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+from torch.utils.data import Dataset
+
+from mmpose.datasets.builder import DATASETS
+
+
+@DATASETS.register_module()
+class MeshAdvDataset(Dataset):
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:8]: ohhh so Adv means adversarial. I thought it was Advanced.

How about using the full name `MeshAdversarialDataset`

Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_adv_dataset.py:8]: OK, I will change it to the full name.

File: mmpose/datasets/datasets/mesh/mesh_base_dataset.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            _iuv_names = [''] * dataset_len
+            has_iuv = 0
+
+        for i in range(len(_imgnames)):
+            newitem = cp.deepcopy(tmpl)
+            newitem['image_file'] = os.path.join(self.img_prefix, _imgnames[i])
+            newitem['scale'] = np.array([_scales[i], _scales[i]])
+            newitem['center'] = _centers[i]
+            newitem['joints_2d'] = _keypoints[i, :, :2]
+            newitem['joints_2d_visible'] = _keypoints[i, :, -1][:, np.newaxis]
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_base_dataset.py:135]: replace all `np.newaxis` with `None`

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    Args:
+        configs (list): List of configs for multiple datasets.
+        partition (list): Sample proportion of multiple datasets.
+    """
+
+    def __init__(self, configs, partition):
+        """Load data from multiple datasets."""
+        self.datasets = [MeshBaseDataset(**cfg) for cfg in configs]
+        self.partition = np.array(partition).cumsum()
+        self.length = max([len(ds) for ds in self.datasets])
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: python supports `max(len(ds) for ds in self.datasets)`

Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: actually if you want to make samples in one sub-dataset are taken with equal probability, the length should not be got by `max`, but by `prod`.

say two datasets with length 2 and 3, and probabilities are 0.5 and 0.5. 
the max length is 3.
the first sample in the first dataset will be sampled more often than the second sample in the first dataset.

Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: however, the length would be too large if we use prod

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    """Mix Dataset for 3D human mesh estimation.
+
+    The dataset combines data from multiple datasets (MeshBaseDataset) and
+    sample the data from different datasets with the provided proportions.
+    The dataset loads raw features and apply specified transforms
+    to return a dict containing the image tensors and other information.
+
+    Args:
+        configs (list): List of configs for multiple datasets.
+        partition (list): Sample proportion of multiple datasets.
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:21]: give an example of the partition. if they are required to be sum to 1, then make this requirement explicit.

Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:21]: and also check it by assert

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py (hunk 15 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: Oh, that's right! Is there other solutions of this problem?

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py (hunk 16 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:21]: OK, I get  it.

File: mmpose/datasets/datasets/mesh/mosh_dataset.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+        self.db = self._get_db(ann_file)
+        self.pipeline = Compose(self.pipeline)
+
+    def _get_db(self, ann_file):
+        """Load dataset."""
+        data = np.load(ann_file)
+        _betas = data['shape'].astype(np.float32)
+        _poses = data['pose'].astype(np.float32)
+        tmpl = dict(
```
Comment (innerlee) [inline mmpose/datasets/datasets/mesh/mosh_dataset.py:39]: maybe revising is not necessary, but it can be shorter

 ```python
gt_db = [dict(pose=p, beta=b) for p, b in zip(_poses, _betas)]
```

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py (hunk 15 above)
Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: what length is proper for the mixed dataset?

Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: It should not be too large. How about use the following code:
```
p = np.random.rand()
for i in range(len(self.datasets)):
    if p <= self.partition[i]:
        index_new = (index + np.random.rand()) * len(self.datasets[i]) / self.length
        index_new = np.round(index_new) % (len(self.datasets[i]))
        return self.datasets[i][index_new]
```

File: mmpose/datasets/datasets/mesh/mosh_dataset.py (hunk 17 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mosh_dataset.py:39]: I prefer to keep the codes here unchanged, because there are many other labels for an item.

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py (hunk 15 above)
Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: looks better

Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py:28]: `int(np.round(index_new)) % ...`
not
`int(np.round(index_new) % ...)`

File: mmpose/datasets/datasets/mesh/mosh_dataset.py (hunk 17 above)
Comment (innerlee) [reply mmpose/datasets/datasets/mesh/mosh_dataset.py:39]: fine

File: mmpose/datasets/datasets/mesh/mesh_mix_dataset.py (hunk 15 above)
Comment (zengwang430521) [reply mmpose/datasets/datasets/mesh/mesh_mix_dataset.py]: I get it.

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
need a following pr to update readme & changelog
